### PR TITLE
feat: add pytorch MACE and UMA

### PR DESCRIPTION
**Pull request by n-gao**
> 🤖 *Auto-generated by Claude. Review and edit as needed.*

# PR Description

Add PyTorch MACE and UMA model support via a universal torch MLFF interface. New `relax_torch` application enables structure relaxation with PyTorch-based potentials alongside existing JAX models.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#178](https://github.com/cusp-ai-oss/kUPS/pull/178) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#169** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/169?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#168** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/168?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#167** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/167?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#166** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/166?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#165** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/165?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#164** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/164?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#155** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/155?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#154** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/154?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#153** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/153?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#152** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/152?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#151** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/151?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#150** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/150?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#149** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/149?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#148** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/148?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#147** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/147?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#146** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/146?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#145** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/145?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#144** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/144?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#143** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/143?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#142** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/142?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#141** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/141?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#138** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/138?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#137** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/137?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#136** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/136?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#135** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/135?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#122** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/122?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#121** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/121?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#134** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/134?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#133** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/133?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#132** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/132?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#131** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/131?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#120** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/120?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/120?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#119** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/119?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#118** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/118?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by felixmusil** [comment]
There are bugs in how the NL is computed that would result in failing tests if there was a 1:1 comparison of the bare-models vs wrapped ones. I discovered this by using the torch_mlip uma model in relaxation resulting in wrong behavior. Here are the issues I found:
+ a bug in the NL #127 
+ it is assumed that the format of the kups NL is the same that UMA expects from `fairchem.core.graph.radius_graph_pbc.radius_graph_pbc_v2`. My understanding is that UMA computes the displacements like `r_j − (r_i + shift·cell)` while kups computes them as `r_j + shift·cell − r_i`  (the direction is reversed). So the forces computed using the kups NL are wrong. In the end I reverted to use the uma internal NL for the moment.
+ Another tangential issue is that the `stress_tensor` is defined in `application.relaxation.data` using the simplest form which is incorrect in most cases. I am adding here a routine to compute the stress from the cell gradients as reference.

```python
def strain_virial_stress(
    cell_vec: Array,
    cell_grad: Array,
    cell_volume: Array,
    positions: Array,
    position_gradients: Array,
    system_indices: Array,
    n_systems: int,
) -> Array:
    """Reconstruct the Cauchy stress tensor from ``∂E/∂h`` and ``∂E/∂r``.

    The naive ``-∂E/∂h / V`` shortcut is dimensionally wrong and silently
    drops the ``hᵀ`` factor and the position-virial term. This helper
    follows the symmetric-strain virial convention used by
    :func:`kups_sim.potential.torch.interface.lattice_gradient_from_virial`:

        W = sym(hᵀ · ∂E/∂h + Σ_b (∂E/∂r_b) ⊗ r_b)
        σ = W / V

    Args:
        cell_vec: Cell vectors ``h`` per system, shape ``(n_systems, 3, 3)``.
        cell_grad: ``∂E/∂h`` per system, shape ``(n_systems, 3, 3)``.
        cell_volume: Cell volume per system, shape ``(n_systems,)``.
        positions: Atomic positions, shape ``(n_atoms, 3)``.
        position_gradients: ``∂E/∂r`` per atom (equal to ``-forces``),
            shape ``(n_atoms, 3)``.
        system_indices: Per-atom integer index into the systems table,
            shape ``(n_atoms,)``.
        n_systems: Number of systems in the batch.

    Returns:
        Cauchy stress tensor per system in eV/Å³, shape
        ``(n_systems, 3, 3)``.
    """
    cell_virial = jnp.einsum("...kj,...kl->...jl", cell_vec, cell_grad)
    per_atom_pos_virial = jnp.einsum("...j,...k->...jk", position_gradients, positions)
    pos_virial = jax.ops.segment_sum(per_atom_pos_virial, system_indices, n_systems)
    virial = cell_virial + pos_virial
    virial = 0.5 * (virial + virial.swapaxes(-1, -2))
    return virial / cell_volume[..., None, None]
```

**Comment by jonkhler** [comment]
@n-gao is this addressed?

**Comment by n-gao** [comment]
#127 is merged, I changed the sign for the offsets for UMA. Regarding the stress tensor, this seems orthogonal since we use the same definition everywhere for now. It would be great if @felixmusil could open an issue that describes what we should implement but that would be a followup.

**Comment by felixmusil** [comment]
I don't think it is orthogonal. If you were to add parity tests checking the outputs of the raw models wrt the kups' packaged ones. This problem would emerge. I guess I am asking for these tests in this PR and such issue would emerge.

**Comment by n-gao** [comment]
In #133, I added end to end tests that the stress we compute aligns with the one from the original models.

**Review by jonkhler — APPROVED** [review]
